rfcs/2020-09-29-4155-aws-s3-source.md

```diff
@@ -0,0 +1,253 @@
+         },
+         "requestParameters":{
+            "sourceIPAddress":"ip-address-where-request-came-from"
+         },
+         "responseElements":{
+            "x-amz-request-id":"Amazon S3 generated request ID",
+            "x-amz-id-2":"Amazon S3 host that processed the request"
+         },
+         "s3":{
+            "s3SchemaVersion":"1.0",
+            "configurationId":"ID found in the bucket notification configuration",
+            "bucket":{
+               "name":"bucket-name",
+               "ownerIdentity":{
+                  "principalId":"Amazon-customer-ID-of-the-bucket-owner"
+               },
+               "arn":"bucket-ARN"
+            },
+            "object":{
+               "key":"object-key",
+               "size":object-size,
+               "eTag":"object eTag",
+               "versionId":"object version if bucket is versioning-enabled, otherwise null",
+               "sequencer": "a string representation of a hexadecimal value used to determine event sequence,
+                   only used with PUTs and DELETEs"
+            }
+         },
+         "glacierEventData": {
+            "restoreEventData": {
+               "lifecycleRestorationExpiryTime": "The time, in ISO-8601 format, for example, 1970-01-01T00:00:00.000Z, of Restore Expiry",
+               "lifecycleRestoreStorageClass": "Source storage class for restore"
+            }
+         }
+      }
+   ]
+}
+```
+
+A S3 client will be initialized when an SQS message is received using the
+`s3.bucket.arn`. It will be cached so we don't need to recerate the client for
+messages referring to the same bucket.
+
+Events will be published for:
+
+* When SQS message is receieved
+  * Number of records in message
+* When SQS message is processed (success or failure)
+  * Number of success / failure / unprocessed records
+* When object is processed
+  * Bytes
+
+These may be revised when the source is actually implemented.
+
+## Doc-level Proposal
+
+We will create a user guide based on
+[AWS's](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/enable-event-notifications.html#s3-event-notification-destinations)
+for configuring bucket notifications to publish object created events to SQS.
+
+The new source configuration will look like:
+
+```toml
+[sources.my_source_id]
+  # General
+  type = "aws_s3" # required
+  compression = "auto" # optional; compression format of the objects; one of: ["auto", "none", "gzip" "lz4" "snappy" "zstd"]; default
+
+  strategy = "sqs" # optional, default, one of ["sqs"]
+
+  sqs.queue_url = "https://sqs.us-east-1.amazonaws.com/1234/test-s3-queue"
+  sqs.poll_secs = 30 # minimum poll interval, optional, default
+  sqs.visibility_timeout_secs = 300 # default visibility timeout for SQS message; if vector does not process the message in this time, the SQS message will be available to be reprocessed
+  sqs.delete_message = true # whether to delete the message after processing; false is useful for debugging; default
```

Comment:
I like the [external tagging](https://serde.rs/enum-representations.html#externally-tagged) pattern you've used here, but it deviates from our normal [internal tagging](https://serde.rs/enum-representations.html#internally-tagged) pattern. For example, the [`fingerprinting` option](https://vector.dev/docs/reference/sources/file/#fingerprinting) uses internal tagging. If we followed that pattern, this would be more like:

```toml
ingestion.strategy = "sqs"
ingestion.queue_url = "https://sqs.us-east-1.amazonaws.com/1234/test-s3-queue"
ingestion.poll_secs = 30
ingestion.visibility_timeout_secs = 300
ingestion.delete_message = true
```

Although, I prefer external tagging since it is clearer and easier to document. If we decide to proceed with this we should open an issue to change the others to be consistent.

Reply:
:thinking: that is a good point. I just went with what felt natural, but I can see if differs from our existing pattern. I'm ok with going either way -- I think being consistent is most important. Failing any other opinions, I'll update this to use the existing internal tagging pattern. I'd probably make it something like:

```toml
ingestion.strategy = "sqs"
ingestion.sql_queue_url = "https://sqs.us-east-1.amazonaws.com/1234/test-s3-queue"
ingestion.sqs_queue_poll_secs = 30
ingestion.sqs_message_visibility_timeout_secs = 300
ingestion.sqs_delete_message = true
```

To make the option names a bit more self-evident.

Tagging @jamtur01 and @lukesteensen for any thoughts.

Reply:
I'm actually coming back around the other way on this one. I think namespacing the config options makes them more readily discoverable and straight-forward rather than needing to read, for each option, whether it applies to my given strategy. I realize this is a shift in the strategy which would necessitate revisiting the other, similar, component config options to deprecate and update them.

I think we can finalize this as part of the PR though since it is easy enough to change.

@binarylogic are you cool with approving this?

Reply:
Yes, the internal tagging looks good to me. I'd prefer that we stay consistent for now.